### PR TITLE
kie-issues#272: Context Menu should always choose the best strategy to render, based on the available space of its container element

### LINKED ISSUE #272: Context Menu should always choose the best strategy to render, based on the available space of its container element.
DMN Boxed expression component context menus may be displayed wrongly in some corner case scenarios. Corner case in this issue means they are invoked at position close to bottom border of the screen and thus some menu items are not visible.

We can illustrate the issue in the picture below.
Screenshot 2023-05-18 184105

There are not visible menu items, becauyse they overlap bottom screen border.

- `Insert below`
- `Delete`
- `Duplicate`
- `Reset`
- `Copy`
- `Cut`
- `Paste`

In similar situations, the context menu should be rendered from the bottom to the top way.

### PR DESCRIPTION
**Pull request by danielzhe**
Closes: https://github.com/kiegroup/kie-issues/issues/272

This PR introduces the logic described bellow to choose where and how to render the context menu, Data Type popup and the drop down in the Data Type popup.

Also there are some videos comparing the old behavior with the new one introduced by this PR.

**Context menu:**
We move the context menu to up in order to be full visible on the screen. The size of the context menu is calculated considering the available items in the menu.

https://github.com/kiegroup/kie-tools/assets/7305741/40a28482-43bc-4582-8b68-826766d7a3fe

**Popup:**
1. We try to position the popup in a way that if user clicks in the "Data type" drop down, there will be available vertical space to show the drop down content
2. We consider the popup with a fixed height of 200px (POPUP_DEFAULT_HEIGHT = 200)
3. We also consider a bottom margin of 200px (POPUP_DROP_DOWN_RESERVED_HEIGHT) for drop down
4. So we position popup in its default place (bottom) or to right if there isn't enough space considering POPUP_DEFAULT_HEIGHT + POPUP_DROP_DOWN_RESERVED_HEIGHT

https://github.com/kiegroup/kie-tools/assets/7305741/8a05ec54-bfd2-49ad-aae8-568421d29636

**Data Type drop down:**
We try to show the entire drop down on to screen limited to 500px (DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT) which we consider the optimal drop down size: is big enough to show a good amount of data but it's not too big.
1. If there is not available space to show the 500px of the drop down, we reduce its size to fit the screen and don't scroll out of the users viewport.
2. We also consider a 46px bottom margin (POPUP_BOTTOM_MARGIN) that is the status bar in the on-line editor. It's ok to use this margin also in VSCode extension.
3. With this margin the bottom border of the drop down is visible, because I found that if is not visible (if we use the entire available space) it may give the feeling that "something is hidden"

https://github.com/kiegroup/kie-tools/assets/7305741/9fef61a9-21a6-45cd-b63d-777ece40e1d6

### CODE REVIEW

**Review by yesamer — COMMENTED** [review]
Thank you @danielzhe, that is a huge improvement for the UX of the editor.
Most of the cases I tested worked nicely.
Just a question, opening the context menu on the most-right will result in showing the menu out of the screen:

Screenshot 2023-05-29 at 14 52 00

Do you think is it possible to manage this case too? Eg. showing the menu on the left of the mouse cursor.

File: packages/boxed-expression-component/src/contextMenu/PopoverMenu/PopoverMenu.tsx [hunk 1]
```diff
@@ -122,14 +126,33 @@ export const PopoverMenu = React.forwardRef(
...
+    const popupPosition = useMemo(() => {
+      if (appendElement) {
+        const yPos = appendElement?.getBoundingClientRect().top ?? 0;
+        const availableHeight = document.documentElement.clientHeight;
+        if (POPUP_DEFAULT_HEIGHT + yPos + POPUP_DROP_DOWN_RESERVED_HEIGHT > availableHeight) {
+          return PopoverPosition.right;
+        }
+      }
+      return PopoverPosition.bottom;
+    }, [appendElement, document.documentElement.clientHeight]);
```
Comment (yesamer) [inline packages/boxed-expression-component/src/contextMenu/PopoverMenu/PopoverMenu.tsx]: @danielzhe Please address the below warnings on useMemo unnecessary deps.

File: packages/boxed-expression-component/src/expressions/ExpressionDefinitionHeaderMenu/DataTypeSelector.tsx [hunk 2]
```diff
@@ -121,6 +127,25 @@ export const DataTypeSelector: React.FunctionComponent<DataTypeSelectorProps> =
...
+      if (
+        DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT <= availableHeight &&
+        DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT + yPos > availableHeight
+      ) {
+        const offset = DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT + yPos - availableHeight;
+        return DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT - offset - POPUP_BOTTOM_MARGIN;
+      }
+    }
+    return DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT;
+  }, [
```
Comment (yesamer) [inline packages/boxed-expression-component/src/expressions/ExpressionDefinitionHeaderMenu/DataTypeSelector.tsx]: @danielzhe Please address the below warnings on useMemo unnecessary deps.

File: packages/boxed-expression-component/src/expressions/ExpressionDefinitionHeaderMenu/ExpressionDefinitionHeaderMenu.tsx [hunk 3]
```diff
@@ -15,8 +15,7 @@
  */
 
 import "./ExpressionDefinitionHeaderMenu.css";
-import * as React from "react";
-import { useCallback, useEffect, useState, useRef } from "react";
+import React, { useCallback, useEffect, useRef, useState } from "react";
```
Comment (yesamer) [inline packages/boxed-expression-component/src/expressions/ExpressionDefinitionHeaderMenu/ExpressionDefinitionHeaderMenu.tsx]: @danielzhe Unnecessary change

**Comment by danielzhe** [comment]
> Thank you @danielzhe, that is a huge improvement for the UX of the editor. Most of the cases I tested worked nicely. Just a question, opening the context menu on the most-right will result in showing the menu out of the screen:
> 
> Screenshot 2023-05-29 at 14 52 00
> Do you think is it possible to manage this case too? Eg. showing the menu on the left of the mouse cursor.

Yes.

Changes applied! It's not exactly at left of the mouse cursor but it's not anymore out of screen. Let me know what you think. We can move it to the left of the cursor if you think this is a best approach.

File: packages/boxed-expression-component/src/expressions/ExpressionDefinitionHeaderMenu/DataTypeSelector.tsx [hunk 4]
```diff
@@ -121,6 +127,20 @@ export const DataTypeSelector: React.FunctionComponent<DataTypeSelectorProps> =
...
+      if (
+        DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT <= availableHeight &&
+        DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT + yPos > availableHeight
+      ) {
+        const offset = DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT + yPos - availableHeight;
+        return DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT - offset - POPUP_BOTTOM_MARGIN;
+      }
+    }
+    return DEFAULT_SELECT_DATA_TYPE_MENU_HEIGHT;
+  }, [selectContainerRef, selectContainerRef.current, selectContainerRef.current?.getBoundingClientRect().top]);
```
Comment (yesamer) [inline packages/boxed-expression-component/src/expressions/ExpressionDefinitionHeaderMenu/DataTypeSelector.tsx]: @danielzhe Please double-check the below dependency warnings.

**Comment by yesamer** [comment]
> > Thank you @danielzhe, that is a huge improvement for the UX of the editor. Most of the cases I tested worked nicely. Just a question, opening the context menu on the most-right will result in showing the menu out of the screen:
> > Screenshot 2023-05-29 at 14 52 00
> > Do you think is it possible to manage this case too? Eg. showing the menu on the left of the mouse cursor.
> 
> Yes.
> 
> Changes applied! It's not exactly at left of the mouse cursor but it's not anymore out of screen. Let me know what you think. We can move it to the left of the cursor if you think this is a best approach.

Thank you, that's much better. Probably showing it on the left would be better for UX, but it's a personal taste, so green lights from me.

**Review by yesamer — APPROVED** [review]
Excellent feature, thank you @danielzhe

**Comment by danielzhe** [comment]
Green. @kiegroup/gatekeepers Merge, please!

**Pull request merged**